yt_dlp/downloader/fragment.py

```diff
@@ -21,6 +23,77 @@
 )
 
 
+class Timer:
+    TIMER_FUNC = time.monotonic
+
+    def __init__(self):
+        self._last_ts = self.TIMER_FUNC()
+
+    def elapsed_seconds(self) -> float:
+        return self.TIMER_FUNC() - self._last_ts
+
+    def has_elapsed(self, seconds: float) -> bool:
+        assert seconds > 0.0
+        elapsed_seconds = self.elapsed_seconds()
+        if elapsed_seconds < seconds:
+            return False
+
+        self._last_ts += elapsed_seconds - elapsed_seconds % seconds
+        return True
+
+
+class SpeedometerMA:
+    TIMER_FUNC = time.monotonic
+    UPDATE_TIMESPAN_S = 1.0
+    AVERAGE_TIMESPAN_S = 5.0
+
+    def __init__(self, initial_bytes=0):
+        self.ts_data = [(self.TIMER_FUNC(), initial_bytes)]
+        self.timer = Timer()
+        self.last_value = None
+
+    def __call__(self, byte_counter):
+        time_now = self.TIMER_FUNC()
+
+        # only append data older than 50ms
+        if time_now - self.ts_data[-1][0] > 0.05:
+            self.ts_data.append((time_now, byte_counter))
+
+        # remove older entries
+        idx = max(0, bisect(self.ts_data, (time_now - self.AVERAGE_TIMESPAN_S, )) - 1)
```

Comment:
Instead of using `max(0, bisect - 1)`, how about using `bisect_left`?

Reply:
This whole bisecting seems weird when you can just use deque instead?

Reply:
`bisect` is used to find the place where we have to threat the values as old values.
Using a `deque` works for storing the data as well, but then you have to call `deque.popleft` in a loop. This might end up faster however and could be considered.

It can be argued that bisecting this problem itself is overkill. We can use simple iteration from the front, since we have 0-2 elements to remove doing that should be about equally as fast.

Reply:
> `bisect` is used to find the place where we have to threat the values as old values. Using a `deque` works for storing the data as well, but then you have to call `deque.popleft` in a loop. 

Nah, just adding a value clears the first value if it's at the storage limit.

Reply:
We want to have an accurate representation over "time", not over datapoints. Datapoints could be very close together, or very far apart (in case of fluctuating internet) which is why we store an arbitrary amount of datapoints based on time

Reply:
It is over time though and it doesn't change anything if you decide to use deque instead.